### DISCUSSION TITLE
What do you think about load speed optimization?

### DISCUSSION THREAD

**Discussion by i-do** [datenstrom/community#289] [Ask a question] [answered]
Hello! I've tested page speed load here:
https://developers.google.com/speed/pagespeed/insights/

There are still many things to optimise. 

You can use Async JavaScript loader & dependency manager to load it faster
https://github.com/globocom/nautilus.js

What do you think?

**Comment by markseuffert** [comment]
Thanks for looking into this.

I remember that when we were looking into speed optimisation the last time, the bottleneck was server side response time. Especially the [Markdown parser](https://github.com/annaesvensson/yellow-markdown) and the [highlight parser](https://github.com/annaesvensson/yellow-highlight) were guilty. Running PHP5 has a negative impact on speed too. Since we use self hosted resources and make extensive use of the browser cache, we didn't see much bottlenecks on the client side response time.

It would be good to get a new overview, before doing optimisations. Can you investigate how Yellow compares speed wise to other content management system? What are the biggest performance bottlenecks for users at the moment? Do you have any problems?

**Comment by dieli** [comment]
I think @markseu is right, that it does not (really) matter for self-hosted sites, where the server response time is the culprit. However, Google search ranks pages based on the speedinsights as well. What I remember from playing around, is to move most of the css and js to the end helps for the score. An integrated solutions (like nautilus) would be very nice!

**Comment by BsNoSi** [comment]
In my opinion Google Insights shows *relative* results. They give you a response to *what may be possible*, that does not reflect the real look and feel of a page. I tested my blog pages a few minutes ago.

> Mobile Needs Work 75 / 100
> Desktop Good 89 / 100

The recommendation is to optimize CSS files. In total we talk about ~ 5k max data savings.
Quite intersting is, that „Mobile“ does not notify response time of server, while „Desktop“ does.

In spite of that: in both cases the page *feels* to be visible instantly. I agree with dieli: in reality, this doesn't matter - these are recommendations for optimization that are disproportionate to the actual benefit.

**Comment by BsNoSi** [comment]
Regarding Nautilus I have to say, that the required jquery has a significant size. There might be an improvement in points but in total it might slow down because of more data in total.

Placement of pages in google results still depends  from content. If you have better matching results, your ranking is nevertheless better than faster pages with worse match.

**Comment by zynicum** [comment]
I just swapped the **Markdown Parser** in Yellow for **Parsedown** and Parsedown-Extra (by  [Emanuil Rusev](http://erusev.com/)). **It is significantly faster and smaller.**
Link: https://github.com/erusev/parsedown
Link Extra: https://github.com/erusev/parsedown-extra
Demo: http://parsedown.org/demo
Demo Extra: http://parsedown.org/extra/
**Unfortunately it is not yet for production** - I still have problems with the blog-plugin (especially the shortcuts). But I am working on it...

Sorry for my English, but I use google-translator

**Comment by BsNoSi** [comment]
I tried the tipp of schulle4u datenstrom/community#254 (commented on 9 Jul) in gallery plugin. This leads to the following Insight-results:

> Mobile Good 91 / 100  (16 points increase)
> Desktop Good 97 / 100  (18 points increase)

This test shows, that the simple change of loading place for plugin assessories can change the ranking significally. Maybe this should be a hint for plugin developer to check, wether the load of script has to be in header or maybe footer is possible. 

In any case, the gallery works smoothly (in my blog...), if the necessary script and CSS is loaded in the footer. That's why I leave it like this:)

**Comment by markseuffert** [comment]
Another suggestion. What we could do in the future is adding a JavaScript module bundler to Yellow. 

At the moment plugins provide multiple JavaScript files and the browser has to load each file. Usually resources will be cached in the browser, but nevertheless each file has to be loaded and updated on the client side. This is where a module bundler comes in. A JavaScript module bundler runs on the server side, it looks at what the API calls "extra HTML data", it will find all included JavaScript files and replaces them with a single bundled JavaScript file.

Instead of a page with multiple resources...

``` html
<!DOCTYPE html>
<html>
<head>
<title>Example page</title>
<script type="text/javascript" src="/media/plugins/gallery.js"></script>
<script type="text/javascript" src="/media/plugins/googlecalendar.js"></script>
<script type="text/javascript" src="/media/plugins/instagram.js"></script>
<script type="text/javascript" src="/media/plugins/slider.js"></script>
<script type="text/javascript" src="/media/plugins/twitter.js"></script>
</head>
<body>
<h1>Hello world</h1>
</body>
</html>
```

...you will get a page with bundled resources:

``` html
<!DOCTYPE html>
<html>
<head>
<title>Example page</title>
<script type="text/javascript" src="/media/plugins/bundle.js"></script>
</head>
<body>
<h1>Hello world</h1>
</body>
</html>
```

And the same thing can be done with CSS. How does this impacts browser performance, I can't say. It might give you a better loading time and a few karma points for search engine optimisations. More research is needed, perhaps someone wants to look into this and meassure alternatives?

**Comment by zynicum** [comment]
@markseu 
A JavaScript module bundler is a nice idea!
BUT: some JavaScript files the browser needs for render the page and some can be loaded after rendering. This behavior can be influenced by "async" and "defer", e.g. <script src="file.js" defer></script> or <script src="file.js" async></script>
[whatwg.org](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-async)

AND: if the website is hosted on HTTP/2, multiple resources can be loaded in parallel!   
[http/2 specification](http://http2.github.io/http2-spec/index.html) ->http2.github.io

Sorry for my English, but I use google-translator

**Comment by BsNoSi** [comment]
I am currently transferring my homepage from a database driven CMS to Yellow. The "beta" looks well and I tested it yesterday  with *Insight*. The results where **really** surprising:

- the mobile result was better than de desktop result
- the mobile result varried over five minutes around 12 (!!) points, when I simply restarted the test.
- Reasons for poor (!) results where quite distracting:
	- five png files, together below 100k, should be kompressed, to achieve a better load of 6%, ~4,20k in total for all images.
	- The css file (6k) schould be better loaded "inline" (for *each* page, mobile user will *love* that…).
	- The fotorama script (loaded from cloudflare) schould be compressed (??)

To display in the wild, what we are talking about: 

- 	19.9 ms Loading
- 	169,9ms Scripting
- 	121.2 ms Rendering
- 	26 ms Painting
- 	109.4 ms Ohter
- 	**Total: 446,40 ms**

After that, the page displays a gallery with five scaled images (1000x300px) and *two external feeds*.

The test today a few minutes ago – without any modifications – leads to 86 points (both "good"). The *look and feel* of the page is *instantlly visible*.

This leads me to the oppinion, that "Insight" is testing potential optimization and not the real performance of a page. The index is – objectivlly speaking – *accademical*.

The test just now leads to 77 (needs work) for mobile and 93 (good) for desktop). Of course without any changes to the page. This test is total nonsense...

A mean question for me is the statement of someone, that Google takes these results as a parameter for site ranking. I cannot believe that.

**Comment by BsNoSi** [comment]
If you want to get some huge scripts only loaded on dedicated pages, you can go further with „petExtra“ in footer snipped. Create a „footerspecial“ snippet and Instead of

```
<?php echo $yellow->page->getExtra("footer") ?>
```
you change the line to

```
<?php echo $yellow->page->getExtra("myspecial") ?>
```
and in respective plugins the code to
```
// Handle page extra HTML data
function onExtra($name)
…
		if($name == "myspecial")
…
```
Then you need a template loading the „footerspecial“ snipped instead of standard footer for the pages, where the scripts of the plugin are needed. 

I tried this for the fotorama-plugin, used on a single page – works fine and is worth to do.

A very upsetting effect of google insight: the better the page is optimised, the worse is impact of slightly changing server response time chances. It varries between 97 to 78 points (!!!), while page "feels" to be instantly available for the user anytime.

**Comment by markseuffert** [comment, marked as answer]
I added a new [plugin to bundle CSS and JavaScript](https://github.com/annaesvensson/yellow-bundle). Let's have a closer look and see what impact a file bundler has on the Datenstrom website. Without it takes 20 requests and 412 ms. With it takes 8 requests and 317 ms. That's the initial loading of resources into the browser cache, consecutive visits are faster. Then it takes 1 request and  <100 ms.

![bundle-screenshot](https://user-images.githubusercontent.com/2429165/44616901-6bdcfa80-a859-11e8-8f1b-9c56a69e28a6.png)

I updated the [highlight plugin](https://github.com/annaesvensson/yellow-highlight). It's included in the next release and comes with a new syntax highlight parser. This makes a big difference, especially if you have a page with a lot of  source code. For example the Datenstrom Yellow API, 850 ms with the old parser, 120 ms with the new parser.

I tested a new [Markdown plugin based on Parsedown Extra](https://github.com/datenstrom/yellow/issues/354). It's not included in the next release. I think it has bugs that would break existing websites. Performance is similar to the current Markdown parser, about 1-10 ms faster per page. Please give it a try and report any problem you find.

Finally, all published plugins have been updated to load JavaScript inside the `<head>` with a `defer` attribute. I recommend that you do the same in your own plugins, unless you have to execute JavaScript at a specific time and place. For more information see: [Efficiently loading JavaScript](https://flaviocopes.com/javascript-async-defer/), [Building the DOM faster](https://hacks.mozilla.org/2017/09/building-the-dom-faster-speculative-parsing-async-defer-and-preload/).

What do you think? How do these optimisations work for you?
